### PR TITLE
use multiple dispatch implementations for buffers

### PR DESCRIPTION
**Pull request by i509VCB**
This simplifies the codebase and removes the need for a Buffer wrapper type that would need conversion all around compositor implementations.

### CODE REVIEW

File: src/wayland/dmabuf/mod.rs [hunk 1]
```diff
@@ -288,15 +290,17 @@ pub enum ImportError {
...
 ///
 /// If the buffer is not managed by the dmabuf handler (whether the buffer is a different kind of buffer,
 /// such as an shm buffer or is not managed by smithay), this function will return an [`UnmanagedResource`]
 /// error.
-pub fn get_dmabuf(buffer: &Buffer) -> Result<Dmabuf, UnmanagedResource> {
-    Ok(buffer.buffer_data::<Dmabuf>()?.ok_or(UnmanagedResource)?.clone())
+///
+/// [`WlBuffer`]: wl_buffer::WlBuffer
+pub fn get_dmabuf(buffer: &wl_buffer::WlBuffer) -> Result<Dmabuf, UnmanagedResource> {
+    Ok(buffer.data::<Dmabuf>().ok_or(UnmanagedResource)?.clone())
```
Comment (vberger) [inline src/wayland/dmabuf/mod.rs:303]:
```suggestion
    buffer.data::<Dmabuf>().cloned().ok_or(UnmanagedResource)
```

**Review by vberger — APPROVED** [review]
One small nit, otherwise LGTM

**Review by PolyMeilex — APPROVED** [review]
This looks a lot cleaner now, nice :ok_hand:

File: src/wayland/dmabuf/mod.rs (hunk 1 above)
Comment (i509VCB) [reply src/wayland/dmabuf/mod.rs:303]: My rationale here is to avoid the clone if the data is not a dmabuf.

Comment (PolyMeilex) [reply src/wayland/dmabuf/mod.rs:303]: Calling Cloned on None will just return None:
```rs
        match self {
            Some(t) => Some(t.clone()),
            None => None,
        }
```
So this is basically free, or is this just stylistic choice?

Comment (i509VCB) [reply src/wayland/dmabuf/mod.rs:303]: Ah okay, so the clone is effectively free then.

**Pull request merged**